nalgebra-sparse/src/io/matrix_market.rs

```diff
@@ -0,0 +1,861 @@
+        } => Err(MMError::from_kind_and_message(
+            MMErrorKind::InvalidHeader,
+            String::from("Pattern matrix can't be hermitian."),
+        )),
+        Typecode {
+            dt: DataType::Pattern,
+            ss: StorageScheme::Skew,
+            ..
+        } => Err(MMError::from_kind_and_message(
+            MMErrorKind::InvalidHeader,
+            String::from("Pattern matrix can't be skew-symmetric."),
+        )),
+        // precheck success
+        _ => Ok(()),
+    }
+}
+
+/// Base trait for matrix market types.
+pub trait MMType: Sized + Clone {
+    /// When matrix is an Integer matrix, it will convert a [i128] number to this type.
+    fn from_i128(i: i128) -> Result<Self, MMError>;
+    /// When matrix is a Real matrix, it will convert a [f64] number to this type.
+    fn from_f64(f: f64) -> Result<Self, MMError>;
+    /// When matrix is a Complx matrix, it will convert a [Complex<f64>] number to this type.
+    fn from_c64(c: Complex<f64>) -> Result<Self, MMError>;
+    /// When matrix is a Pattern matrix, it will convert a unit type [unit] to this type.
+    fn from_pattern(p: ()) -> Result<Self, MMError>;
+    /// When matrix is a Skew-symmetric matrix, it will convert itself to its negative.
+    fn negative(self) -> Result<Self, MMError>;
+    /// When matrix is a Hermitian matrix, it will convert itself to its conjugate.
+    fn conjugate(self) -> Result<Self, MMError>;
+}
+/// Implement MMType for primitive integer types.
+macro_rules! mm_int_impl {
+    ($T:ty) => {
+        impl MMType for $T {
+            #[inline]
+            fn from_i128(i: i128) -> Result<Self, MMError> {
+                Ok(i as Self)
+            }
+            #[inline]
+            fn from_f64(_f: f64) -> Result<Self, MMError> {
+                Err(MMError::from_kind_and_message(
+                    MMErrorKind::TypeUnmatched,
+                    format!("Int type can't parse from f64"),
+                ))
+            }
+            #[inline]
+            fn from_c64(_c: Complex<f64>) -> Result<Self, MMError> {
+                Err(MMError::from_kind_and_message(
+                    MMErrorKind::TypeUnmatched,
+                    format!("Int type can't parse from Complex<f64>"),
+                ))
+            }
+            #[inline]
+            fn from_pattern(_p: ()) -> Result<Self, MMError> {
+                Err(MMError::from_kind_and_message(
+                    MMErrorKind::TypeUnmatched,
+                    format!("Int type can't parse from ()"),
+                ))
+            }
+            #[inline]
+            fn conjugate(self) -> Result<Self, MMError> {
+                Err(MMError::from_kind_and_message(
+                    MMErrorKind::TypeUnmatched,
+                    format!("Int type has no conjugate"),
+                ))
+            }
+            #[inline]
+            fn negative(self) -> Result<Self, MMError>
+            where
+                Self: Neg<Output = Self>,
```

Comment:
since you're just doing this in a macro, you don't need the additional trait bound I think